File: 字母的移位词分组.cpp


#include <iostream>
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
	vector<vector<string>> groupAnagrams(vector<string> &strs) {
		vector<vector<string>> res;
		unordered_map<string, vector<string>> m;
		for(string s:strs) {
			string t = s;
			sort(t.begin(), t.end());
			m[t].push_back(s);
		}

		for(auto n:m) {
			res.push_back(n.second);
		}
		return res;
	}
};

int main(int argc, char* argv[]) {

	vector<string> res{"eat", "tea", "tan", "ate", "nat", "bat"};
	vector<vector<string>> temp;
	Solution s;
	temp = s.groupAnagrams(res);
	for(auto it:temp) {
		for(auto j:it) {
			cout<<j<<" ";
		}
		cout<<endl;
	}

	return 0;
}

File: string/字母的移位词分组.java


class Solution {
    public static List<List<String>> groupAnagrams(String[] strs) {
        //哈希表，排好序的单词 -> 该单词对应的单词集合
        Map<String, List<String>> groups = new HashMap<>();

        for (String str : strs) {
            char[] charArray = str.toCharArray();

            //计算每个单词映射的排好序的单词
            Arrays.sort(charArray);
            String newStr = new String(charArray);

            groups.putIfAbsent(newStr, new LinkedList<>());
            groups.get(newStr).add(str);
        }
        return new LinkedList<>(groups.values());
    }

    public static void main(String[] args) {

        String[] strs = {"eat", "tea", "tan", "ate", "nat", "bat"};
        List<List<String>> list2 = new ArrayList<>();

        list2 = groupAnagrams(strs);
        for (var i : list2) {
            for (var j : i) {
                System.out.print(j + " ");
            }
            System.out.println();
        }
    }
}
